lib/PuppeteerSharp/CDPSession.cs

```diff
@@ -184,22 +184,32 @@ internal void OnMessage(ConnectionResponse obj)
         {
             var id = obj.Id;
 
-            if (id.HasValue && _callbacks.TryRemove(id.Value, out var callback))
+            if (id.HasValue)
             {
-                if (obj.Error != null)
+                if (_callbacks.TryRemove(id.Value, out var callback))
                 {
-                    callback.TaskWrapper.TrySetException(new MessageException(callback, obj.Error));
-                }
-                else
-                {
-                    callback.TaskWrapper.TrySetResult(obj.Result);
+                    // This is a response to a SendAsync. Handle the callback async, since (a) we can, and (b)
+                    // to avoid a situation where the continuation tries to call SendAsync again thus creating
+                    // a deadlock.
+                    Task.Run(() =>
+                    {
+                        if (obj.Error != null)
+                        {
+                            callback.TaskWrapper.TrySetException(new MessageException(callback, obj.Error));
+                        }
+                        else
+                        {
+                            callback.TaskWrapper.TrySetResult(obj.Result);
+                        }
+                    });
```

Comment:
My main concern here is the ASP.NET Full Framework Scenario. Could you test [this project](https://github.com/hardkoded/puppeteer-sharp/tree/master/samples/PupppeterSharpAspNetFrameworkSample) using these changes?

Reply:
I'm not totally sure if I set up the sample correctly, but got it to run locally with IIS Express (since I don't have full IIS installed locally), and if I open in my own browser /PupppeterSharpAspNetFrameworkSample/api/test, I get back `"\"bar\""`. I set breakpoints in the debugger to at least make sure my `Task.Run()` tasks were being called and they are.

Is there anything else you can think of that I should test?

Reply:
@kblok any luck on testing this in full IIS?